README.md

```diff
@@ -1,3 +1,262 @@
+- [Preliminary Measurements](#preliminary-measurements)
+- [Contributing](#contributing)
+- [Contributors](#contributors)
+- [License: Apache 2.0](#license-apache-20)
+- [Future](#future)
+
+###Motivation
+
+---
+Angular 2 (which is written in TypeScript) currently sits at around 135kb after being [Uglified](https://github.com/mishoo/UglifyJS) and compressed through GZIP. In comparison, Angular 1.4 is about 50kb, minified and compressed.
+
+A smaller bundle size means that less data needs to be transferred and loaded by the browser. This contributes to a better user experience.
+
+The impetus for this tool was to reduce the code size of the Angular 2 bundle, but TS-Minify is meant to be a generic tool that can be used on programs written in TypeScript.
+
+
+###What it Does
+
+---
+To achieve code size reduction, TS-Minify uses the idea of property renaming: take a TypeScript source file and rename properties to shorter property names, then re-emit the file as valid TypeScript.
+
+![TS-Minify role in minification pipeline](http://i.imgur.com/7iH4RyF.png) <sub>This diagram demonstrates TS-Minify's role in the intended minification pipeline.</sub>
+
+TS-Minify *only*  targets property renaming. Minification tactics like whitespace removal, dead code removal, variable name mangling, etc. are taken care of by tools such as [UglifyJS](https://github.com/mishoo/UglifyJS). TS-Minify specifically targets property renaming since it is something that is difficult to achieve *safely* without type information. As such, TS-Minify requires a program to be correctly and throughly typed, otherwise, unwanted renaming may occur. 
+
+###How Does it Work? The TL;DR
+
+---
+The TypeScript Compiler API is utilized in order to access the Abstract Syntax Tree of a TypeScript source file. Identifiers that might be considered properties (such as identifiers in property declarations, property access expressions, method names in method declarations, etc.) are renamed to short names such as `a`, `b`, etc. after determining their renaming eligibility. A “renaming” global map is created with mappings from the original property name to the new generated property name so that re-namings are kept consistent between properties with the same names.
+
+```javascript
+{ maybeHandleCall: g }
+{ reportMissingType: h }
+{ getHandler: i }
+{ handlePropertyAccess: j }
+{ emitExtraImports: k }
+{ emit: l }
+{ visitTypeName: m }
+```
+<sub>An example of some mappings from the original property name to the shorter, generated property name.</sub>
+
+The renaming eligibility of a property takes several factors into consideration:
+
+- Does the property belong to an object declared in an external file ([`.d.ts file`](http://definitelytyped.org/))?
+- Does the property belong to a [standard built-in object](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects)?
+- Does the property name belong to the DOM?
+- Does the property belong to an object being exported?
```

Comment:
I think we don't really tackle per-file minification mode, thus don't care about es6 module exports.